File: spring-webflux/src/main/java/com/example/webflux/domain/CpuService.java
package com.example.webflux.domain;

import com.example.webflux.observability.EventLog;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.concurrent.atomic.AtomicBoolean;

public interface CpuService {
    Mono<Void> run(String req);
}

@Service
class CpuServiceImpl implements CpuService {

    private static final int DURATION_SECONDS = 15;
    private static final long YIELD_INTERVAL_MS =
        Long.parseLong(System.getenv().getOrDefault("YIELD_INTERVAL_MS", "100"));

    @Override
    public Mono<Void> run(String req) {
        return Mono.<Void>create(sink -> {
            AtomicBoolean cancelled = new AtomicBoolean(false);
            sink.onCancel(() -> cancelled.set(true));

            EventLog.log(req, EventLog.Stage.SERVICE, "stage_entered");

            long start = System.currentTimeMillis();
            long iters = 0;
            long primes = 0;
            long candidate = 2;
            long lastCheckMs = start;

            EventLog.log(req, EventLog.Stage.SERVICE, "work_started");

            while (System.currentTimeMillis() - start < DURATION_SECONDS * 1000L) {
                long now = System.currentTimeMillis();
                if (now - lastCheckMs >= YIELD_INTERVAL_MS) {
                    lastCheckMs = now;
                    if (cancelled.get()) {
                        long elapsed = now - start;
                        EventLog.log(req, EventLog.Stage.SERVICE, "cancellation_detected",
                                "iters=" + iters + " primes=" + primes + " elapsed_ms=" + elapsed);
                        return;
                    }
                }

                if (isPrime(candidate)) primes++;
                candidate++;
                iters++;
            }

            long elapsed = System.currentTimeMillis() - start;
            EventLog.log(req, EventLog.Stage.SERVICE, "work_completed",
                    "iters=" + iters + " primes=" + primes + " elapsed_ms=" + elapsed);
            sink.success();
        }).subscribeOn(SchedulerConfig.CPU);
    }

    private static boolean isPrime(long n) {
        if (n < 2) return false;
        if (n == 2) return true;
        if (n % 2 == 0) return false;
        for (long i = 3; i * i <= n; i += 2)
            if (n % i == 0) return false;
        return true;
    }
}

File: aspnet/src/Domain/CpuService.cs
using System.Diagnostics;
using static src.Observability.EventLog;

namespace src.Domain;

public interface ICpuService
{
    Task RunAsync(string req, CancellationToken ct);
}

public class CpuService : ICpuService
{
    internal static readonly int DurationSeconds =
        int.TryParse(Environment.GetEnvironmentVariable("CPU_DURATION_S"), out var d) ? d : 15;
    internal static readonly long YieldIntervalMs =
        long.TryParse(Environment.GetEnvironmentVariable("YIELD_INTERVAL_MS"), out var v) ? v : 100;

    public async Task RunAsync(string req, CancellationToken ct)
    {
        Log(req, Stage.Service, "stage_entered");

        var sw = Stopwatch.StartNew();
        long iters = 0;
        long primes = 0;
        long candidate = 2;
        long lastYieldMs = 0;

        Log(req, Stage.Service, "work_started");

        while (sw.Elapsed.TotalSeconds < DurationSeconds)
        {
            if (ct.IsCancellationRequested)
            {
                Log(req, Stage.Service, "cancellation_detected",
                    $"iters={iters} primes={primes} elapsed_ms={sw.ElapsedMilliseconds}");
                throw new OperationCanceledException(ct);
            }

            if (IsPrime(candidate)) primes++;
            candidate++;
            iters++;

            if (sw.ElapsedMilliseconds - lastYieldMs >= YieldIntervalMs)
            {
                lastYieldMs = sw.ElapsedMilliseconds;
                await Task.Yield();
            }
        }

        Log(req, Stage.Service, "work_completed",
            $"iters={iters} primes={primes} elapsed_ms={sw.ElapsedMilliseconds}");
    }

    private static bool IsPrime(long n)
    {
        if (n < 2) return false;
        if (n == 2) return true;
        if (n % 2 == 0) return false;
        for (long i = 3; i * i <= n; i += 2)
            if (n % i == 0) return false;
        return true;
    }
}
